Reapply "splice.nix: make pkgs splicedPackages"

wolfgangwalther opened the pull request:
This reapplies #349316, which was tried in #362499, but then abandoned.

Does https://github.com/NixOS/nixpkgs/pull/349316#issuecomment-2478607271, too, but not https://github.com/NixOS/nixpkgs/pull/349316#discussion_r1826155762. The latter still causes me some infrec headache for darwin cross.

Dropping `stdenv/native` along the way, which was broken right now anyway - and causes infinite recursions through these `nativeTools` & co. arguments.

## Things done

- [x] Basic tests work.
- [ ] Do more sophisticated testing.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Artturin committed:
splice.nix: make `pkgs` `splicedPackages` when required

This will make `pkgs` used in `callPackage`, and `pkgsCross.X.pkgs` have
packages with `__spliced`.

https://www.github.com/NixOS/nixpkgs/blob/3029741718f4c765fbc5ebf76bea3d6c8ff15fe5/pkgs/development/interpreters/python/passthrufun.nix#L37

https://www.github.com/NixOS/nixpkgs/blob/d2bd9a39dec88eddd5c192abee69939e67f43d12/pkgs/top-level/python-packages.nix#L10720

```
nix-repl>
pkgsCross.aarch64-multiplatform.python3Packages.protobuf4.protobuf.__spliced
error:
       … while evaluating the attribute
'aarch64-multiplatform.python3Packages.protobuf4.protobuf.__spliced'
         at
/home/artturin/nixgits/my-nixpkgs/.worktree/1/pkgs/development/python-modules/protobuf/4.nix:119:13:
          118|   passthru = {
          119|     inherit protobuf;
             |             ^
          120|   };

       error: attribute '__spliced' missing
       at «string»:1:1:
            1|
pkgsCross.aarch64-multiplatform.python3Packages.protobuf4.protobuf.__spliced
             | ^
```

to

```
nix-repl>
pkgsCross.aarch64-multiplatform.python3Packages.protobuf4.protobuf.__spliced
{
  buildBuild = «derivation
/nix/store/s7da5mfvx4h1n86j78knaj9cprglxqz6-protobuf-25.4.drv»;
  buildHost = «derivation
/nix/store/s7da5mfvx4h1n86j78knaj9cprglxqz6-protobuf-25.4.drv»;
  buildTarget = «repeated»;
  hostHost = «derivation
/nix/store/mszvybzs4zxh43awyrjnybsfcb265n9r-protobuf-aarch64-unknown-linux-gnu-25.4.drv»;
  hostTarget = «repeated»;
}
```

Co-authored-by: Wolfgang Walther <walther@technowledgy.de>

Artturin committed:
treewide: remove `= __splicedPackages`

Now that `pkgs` is `__splicedPackages` on cross we can remove these
variables I set.

Assignments like `patchutils = pkgs.patchutils_0_3_3;` in
`all-packages.nix` cannot be removed because the `pkgs`
in `patchutils = pkgs.patchutils_0_3_3;` in `all-packages.nix` is coming from
this `pkgs:` https://www.github.com/NixOS/nixpkgs/blob/b6e486730fc875ec79a3dea0f1f46eaf9f6ffa9d/pkgs/top-level/all-packages.nix#L9
instead of the `pkgs` in `with pkgs;`

wolfgangwalther committed:
all-packages.nix: remove unused headers

wolfgangwalther committed:
top-level: return __splicedPackages

wolfgangwalther commented:
> > Dropping `stdenv/native` along the way, which was broken right now anyway - and causes infinite recursions through these `nativeTools` & co. arguments.
> 
> The changes here wouldn’t cause that without the `__splicedPackages` in `all-packages.nix`, right?

That's what I thought at first, but it doesn't seem to be the case, no, at least when I tried this morning. OTOH, I had received confusingly different results from time to time, so I'm not really trusting myself to always hold this correctly. I will test and confirm it again, but my current idea is that this is indeed already required for the actual reapply.

szlend commented on lib/customisation.nix:
```diff
@@ -726,7 +726,7 @@ rec {
       };
       spliced = extra spliced0 // spliced0 // keep self;
-      self = f self // {
+      self = f spliced0 // {
```
It's not exactly what I had in mind, but that indeed looks like a problem yeah.

szlend replied:
It doesn't feel to me like this can be avoided with the way splicing works though.

What if we kept `self = f self // ...` and instead made `makeScopeWithSplicing'` return `spliced0`? It's not as simple as that, but it would sort of mimic what was done to `pkgs` where `pkgs` was redefined as `pkgs.__splicedPackages`.

wolfgangwalther commented:
> I tested again: The reapply itself works in all my basic tests. It's the `lib.customisation: make self spliced in scopes` commit that causes `stdenv.cc.nativeTools` infrec for `pkgsLLVM` now, which makes the commit to drop stdenv/native required here.

I also looked into the eval failure by CI in `nixComponents_...` - this is caused by the same commit. To reduce scope, I removed the commit about scopes for now.

I'll bring the drop for stdenv/native back with the scope commit in a separate PR.

wolfgangwalther committed:
systems/examples: remove x86_64-unknown-redox

relibc itself has been removed in 2024 already.

wolfgangwalther committed:
uclibc: bring back alias

This was removed recently when cleaning up aliases, but it breaks
`pkgsCross.ben_nanonote`.

wolfgangwalther commented on pkgs/top-level/splice.nix:
```diff
@@ -173,6 +173,8 @@ in
 
   newScope = extra: lib.callPackageWith (pkgsForCall // extra);
 
+  pkgs = if actuallySplice then splicedPackages // { recurseForDerivations = false; } else pkgs;
```
It's not a `let` but just a regular attrset definition here - without `rec`. The `pkgs` here is from line 17, `lib: pkgs: actuallySplice:`, i.e. function args.

philiptaron reviewed (commented):
I read through this commit-by-commit. Some questions were straightforward to answer, but a couple places could use bigger comments explaining what and why.

wolfgangwalther commented on pkgs/top-level/aliases.nix:
```diff
@@ -254,6 +254,7 @@ mapAliases {
   elasticsearch7Plugins = elasticsearchPlugins; # preserve, reason: until v8
   fetchFromGithub = throw "You meant fetchFromGitHub, with a capital H"; # preserve, reason: common typo
   fuse2fs = if stdenv.hostPlatform.isLinux then e2fsprogs.fuse2fs else null; # Added 2022-03-27 preserve, reason: convenience, arch has a package named fuse2fs too.
+  uclibc = uclibc-ng; # preserve, because uclibc-ng can't be used in config string
```
No, it's the only one of our examples that breaks this way. Look at this snippet:

https://github.com/NixOS/nixpkgs/blob/ddf56acbeb53825b79b82c13a657806774bfd137/lib/systems/examples.nix#L107-L109

The libc is appended as `-uclibc` to the config string. If we append `uclibc-ng`, this will break, because the parser for that string splits on `-` and expects a certain number of fields.

philiptaron commented:
I would very much shy away from providing anything other than best wishes for that scenario.

wolfgangwalther commented:
In brought back the additional changes in the following two draft PRs:
- #458108
- #458111

These don't work, yet, any help appreciated.

link2xt commented:
This change broke naersk example of cross-compilation from Linux to Windows according to `git bisect`:
https://github.com/nix-community/naersk/issues/371
